Mage/src/main/java/mage/util/CardUtil.java

```diff
@@ -1791,7 +1791,7 @@ public static <T> T getSourceCostsTag(Game game, Ability source, String tag, T d
             if (value == null) {
                 throw new IllegalStateException("Wrong code usage: Costs tag " + tag + " has value stored of type null but is trying to be read. Use checkSourceCostsTagExists");
             }
-            if (value.getClass() != defaultValue.getClass()) {
+            if (defaultValue != null && value.getClass() != defaultValue.getClass()) {
```

Comment:
@ssk97 could you review the usage of cost tags in this PR?

Reply:
I personally don't like storing both nulls and player ID under the same tag. It works, but I think I would have used two separate tags. At the very least I strongly suggest adding a comment by the definition of `GIFT_ACTIVATION_VALUE_KEY` explaining that it stores `null` if activated but the opponent is not yet selected, and then the UUID of the player chosen.

The change here removes the precise type-checking, but since there's then the typecast immediately afterwards it should be fine. That precise type-checking isn't truly necessary I suppose, the requirement of it being castable is enough.

Reply:
I honestly didn't really know what to do here and was going off of inference, is there anything I should change about it?

Reply:
There's things that I personally would've done differently, but the only thing I would say definitely should be done is adding a comment mentioning that the tag is used both with null and with the opponent UUID.